src/charm.py

```diff
@@ -4,101 +4,424 @@
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
 
-    def _on_httpbin_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API.
+        self._port = 80
+        self._diagnostics_port = 8082
+
+        self.service_patch = KubernetesServicePatch(
+            charm=self,
+            service_type="LoadBalancer",
+            ports=[(f"{self.app.name}", self._port)],
+        )
+
+        self.metrics_endpoint = MetricsEndpointProvider(
+            self,
+            jobs=[
+                {
+                    "static_configs": [{"targets": [f"*:{self._diagnostics_port}"]}],
+                },
+            ],
+        )
+
+        self.framework.observe(self.on.traefik_pebble_ready, self._on_traefik_pebble_ready)
+
+        # TODO Need to ensure K8s service has LoadBalancer IP before processing proxy requests?
+
+        self.framework.observe(self.on["ingress"].relation_created, self._handle_ingress_change)
+        self.framework.observe(self.on["ingress"].relation_joined, self._handle_ingress_change)
+        self.framework.observe(self.on["ingress"].relation_changed, self._handle_ingress_change)
+        self.framework.observe(self.on["ingress"].relation_departed, self._handle_ingress_change)
+        self.framework.observe(self.on["ingress"].relation_broken, self._handle_ingress_broken)
+
+    def _on_traefik_pebble_ready(self, event):
+        # Ensure the required basic configurations and folders exist
+        container = self.unit.get_container(_TRAEFIK_CONTAINER_NAME)
+
+        # TODO Handle case the relation events are deferred and this hook has not
+        # run yet when the pebble event queueing triggers the previously deferred
+        # events.
+        if not container.can_connect():
+            self.unit.status = WaitingStatus(
+                f"container '{_TRAEFIK_CONTAINER_NAME}' not yet ready"
+            )
+            return
+
+        # TODO Use the Traefik user and group?
+        container.make_dir(path="/etc/traefik", make_parents=True)
+        container.make_dir(path=_TRAEFIK_INGRESS_CONFIGURATIONS_DIRECTORY, make_parents=True)
+
+        # Since pebble ready will also occur after a pod churn, but we store the
+        # configuration files on a storage volume that survives the pod churn, before
+        # we start traefik we clean up all Juju-generated config files to avoid spurious
+        # routes.
+        for ingress_relation_configuration_file in container.list_files(
+            path=_TRAEFIK_INGRESS_CONFIGURATIONS_DIRECTORY,
+            pattern="juju_*.yaml"
+        ):
+            container.remove_path(ingress_relation_configuration_file.path)
+            logger.debug(f"Deleted orphaned ingress configuration file: {ingress_relation_configuration_file.path}")
+
+        # TODO Disable static config BS with telemetry and check new version
+        basic_configurations = yaml.dump(
```

Comment:
```suggestion
        static_configuration = yaml.dump(
```

Might be nice to rename this so it's consistent with Traefik's docs